File: zen/color.py
import os
import sys
import re

_TERMUX = 'TERMUX_VERSION' in os.environ


class Color:
    def __init__(self):
        self._enabled = sys.stdout.isatty() and 'NO_COLOR' not in os.environ

    def enable(self, val=True):
        self._enabled = val

    @property
    def enabled(self):
        return self._enabled

    def _fmt(self, code, text):
        if text is None:
            return f'\033[{code}m' if self._enabled else ''
        if not self._enabled:
            return str(text)
        return f'\033[{code}m{text}\033[0m'

    def _fmt_pair(self, code1, code2, text):
        if text is None:
            return f'\033[{code1};{code2}m' if self._enabled else ''
        if not self._enabled:
            return str(text)
        return f'\033[{code1};{code2}m{text}\033[0m'

    def prompt(self, text):
        if not self._enabled:
            return str(text)
        if _TERMUX:
            return text
        return f'\001\033[0m\002{text}\001\033[0m\002'

    def reset(self, text=None):
        return self._fmt(0, text)

    def bold(self, text=None):
        return self._fmt(1, text)

    def dim(self, text=None):
        return self._fmt(2, text)

    def italic(self, text=None):
        return self._fmt(3, text)

    def underline(self, text=None):
        return self._fmt(4, text)

    def blink(self, text=None):
        return self._fmt(5, text)

    def reverse(self, text=None):
        return self._fmt(7, text)

    def hidden(self, text=None):
        return self._fmt(8, text)

    def strike(self, text=None):
        return self._fmt(9, text)

    def black(self, text=None):
        return self._fmt(30, text)

    def red(self, text=None):
        return self._fmt(31, text)

    def green(self, text=None):
        return self._fmt(32, text)

    def yellow(self, text=None):
        return self._fmt(33, text)

    def blue(self, text=None):
        return self._fmt(34, text)

    def magenta(self, text=None):
        return self._fmt(35, text)

    def cyan(self, text=None):
        return self._fmt(36, text)

    def white(self, text=None):
        return self._fmt(37, text)

    def bright_black(self, text=None):
        return self._fmt(90, text)

    def bright_red(self, text=None):
        return self._fmt(91, text)

    def bright_green(self, text=None):
        return self._fmt(92, text)

    def bright_yellow(self, text=None):
        return self._fmt(93, text)

    def bright_blue(self, text=None):
        return self._fmt(94, text)

    def bright_magenta(self, text=None):
        return self._fmt(95, text)

    def bright_cyan(self, text=None):
        return self._fmt(96, text)

    def bright_white(self, text=None):
        return self._fmt(97, text)

    def rgb(self, r, g, b, text=None):
        if not self._enabled:
            return str(text) if text is not None else ''
        if text is not None:
            return f'\033[38;2;{int(r)};{int(g)};{int(b)}m{text}\033[0m'
        return f'\033[38;2;{int(r)};{int(g)};{int(b)}m'

    def hex(self, h, text=None):
        h = str(h).lstrip('#')
        if len(h) == 3:
            h = ''.join(c * 2 for c in h)
        r, g, b = int(h[0:2], 16), int(h[2:4], 16), int(h[4:6], 16)
        return self.rgb(r, g, b, text)

    def strip(self, text):
        return re.sub(r'\033\[[0-9;]*m', '', str(text))


color = Color()

File: src/color.rs
//! Zen `color` module.

use crate::runtime::{Vm, Value};
use std::sync::Arc;

pub fn init_color_module(vm: &mut Vm) {
let color = Value::Dict(Arc::new(ahash::AHashMap::from([
    ("reset".into(), Value::String("\x1b[0m".into())),
    ("bold".into(), Value::NativeFunction("color_style_bold".into())),
    ("dim".into(), Value::NativeFunction("color_style_dim".into())),
    ("italic".into(), Value::NativeFunction("color_style_italic".into())),
    ("underline".into(), Value::NativeFunction("color_style_underline".into())),
    ("blink".into(), Value::NativeFunction("color_style_blink".into())),
    ("reverse".into(), Value::NativeFunction("color_style_reverse".into())),
    ("hidden".into(), Value::NativeFunction("color_style_hidden".into())),
    ("strike".into(), Value::NativeFunction("color_style_strike".into())),
    ("rgb".into(), Value::NativeFunction("color_rgb".into())),
    ("bg_rgb".into(), Value::NativeFunction("color_bg_rgb".into())),
    ("hex".into(), Value::NativeFunction("color_hex".into())),
    ("strip".into(), Value::NativeFunction("color_strip".into())),
    ("black".into(), Value::NativeFunction("color_fg_black".into())),
    ("red".into(), Value::NativeFunction("color_fg_red".into())),
    ("green".into(), Value::NativeFunction("color_fg_green".into())),
    ("yellow".into(), Value::NativeFunction("color_fg_yellow".into())),
    ("blue".into(), Value::NativeFunction("color_fg_blue".into())),
    ("magenta".into(), Value::NativeFunction("color_fg_magenta".into())),
    ("cyan".into(), Value::NativeFunction("color_fg_cyan".into())),
    ("white".into(), Value::NativeFunction("color_fg_white".into())),
    ("bg_black".into(), Value::NativeFunction("color_bg_black".into())),
    ("bg_red".into(), Value::NativeFunction("color_bg_red".into())),
    ("bg_green".into(), Value::NativeFunction("color_bg_green".into())),
    ("bg_yellow".into(), Value::NativeFunction("color_bg_yellow".into())),
    ("bg_blue".into(), Value::NativeFunction("color_bg_blue".into())),
    ("bg_magenta".into(), Value::NativeFunction("color_bg_magenta".into())),
    ("bg_cyan".into(), Value::NativeFunction("color_bg_cyan".into())),
    ("bg_white".into(), Value::NativeFunction("color_bg_white".into())),
    ("bright_black".into(), Value::NativeFunction("color_fg_bright_black".into())),
    ("bright_red".into(), Value::NativeFunction("color_fg_bright_red".into())),
    ("bright_green".into(), Value::NativeFunction("color_fg_bright_green".into())),
    ("bright_yellow".into(), Value::NativeFunction("color_fg_bright_yellow".into())),
    ("bright_blue".into(), Value::NativeFunction("color_fg_bright_blue".into())),
    ("bright_magenta".into(), Value::NativeFunction("color_fg_bright_magenta".into())),
    ("bright_cyan".into(), Value::NativeFunction("color_fg_bright_cyan".into())),
    ("bright_white".into(), Value::NativeFunction("color_fg_bright_white".into())),
])));
vm.vars.insert("color".into(), color);
}
